Add support for hybrid environments

jpculp opened the pull request:
**Issue number:**

N/A

**Description of changes:**

This change allows users running on-premises or in a hybrid environment to register with SSM as a managed-instance. SSM's local state directory (`/var/lib/amazon/ssm`) has moved to a persistent directory in `/.bottlerocket/host-containers/current/`. Activations parameters are parsed from a base64-encoded JSON block in user-data.

Example JSON:
```
{
  "ssm":{
    "activation_id":"foo",
    "activation_code":"bar",
    "region":"us-west-2",
  }
}
```

Also bumped the SSM Agent version to [3.0.882.0](https://github.com/aws/amazon-ssm-agent/releases/tag/3.0.882.0).

And bumped container version to **v0.5.0**.

**Testing done:**

- Created an ssm hybrid activation.
- Launched `aws-ecs-1` ami with added `[settings.host-containers.control] user-data`.
- Instance connected to ecs cluster.
- Test task deployed successfully.
- Check AWS Systems Manager to see if my activation was used.
- Fetched instance-id from AWS Systems Manager and verified it began with `mi-` prefix.
- Connected to control container via ssm session.
- Verified that `host-containers.control.user-data` contained a base64-encoded block.
- Disabled the `control` container and enabled a `current` container.
- Verified that /.bottlerocket/host-containers/current/ contained an ssm directory.
- Enabled and launched admin container.
- Connected to admin container via ssh.
- Ran `sudo sheltie` to verify root shell was still available.
- Checked for failed systemd units.

- Launched additional instance with no user data and verified everything still works.
- Launched additional instances with purposefully malformed user data detailed below.

**Some possible error messages:**
<details><summary>Bad Region: us-weast-2</summary>
<p>

```
[   25.190387] host-ctr[447]: 2021-03-31 04:38:44 ERROR [processRegistration @ agent_parser.go.128] Registration failed due to error registering the instance with AWS SSM. RequestError: send request failed
[   25.192567] host-ctr[447]: caused by: Post "https://ssm.us-weast-2.amazonaws.com/": dial tcp: lookup ssm.us-weast-2.amazonaws.com on 172.31.0.2:53: no such host
[   25.212203] host-ctr[447]: Failed to register with AWS Systems Manager (SSM)
```

</p>
</details>

<details><summary>Bad Activation ID: character missing when copy/pasting</summary>
<p>

```
[   41.240253] host-ctr[445]: 2021-03-31 04:39:04 ERROR [processRegistration @ agent_parser.go.128] Registration failed due to error registering the instance with AWS SSM. InvalidParameter: 1 validation error(s) found.
[   41.244048] host-ctr[445]: - minimum field size of 36, RegisterManagedInstanceInput.ActivationId.
[   41.245795] host-ctr[445]: Failed to register with AWS Systems Manager (SSM)
```

</p>
</details>

<details><summary>Empty Element: "activation-code":""</summary>
<p>

```
[   13.964770] host-ctr[441]: Failed to fetch value for .["ssm"]["activation-code"] from /.bottlerocket/host-containers/current/user-data
```

</p>
</details>

<details><summary>Missing Element: no region specified</summary>
<p>

```
[   13.195573] host-ctr[445]: Failed to fetch value for .["ssm"]["region"] from /.bottlerocket/host-containers/current/user-data
```

</p>
</details>

**SSM Agent changes since v0.4.2:**

<details><summary>3.0.882.0</summary>
<p>

- Added jitter to first control channel call
- Added dedicated folder for plugins
- Added option to overwrite corrupt shared credentials

</p>
</details>

<details><summary>3.0.854.0</summary>
<p>

- Added $HOME env variable for root user when runAsElevated is true in session
- Added CREAD flag in serial port control flags on linux
- Added PlatformName and PlatformVersion as env variables for aws:runShellScript
- Added support for macOS updater
- Added v2.2 document support in updater
- Added defer recover statements
- Fixed inventory error log when dpkg is not available
- Fixed ssm-cli logging to stdout
- Removed consideration of unimportant error codes in service side
- Updated ec2 credential caching time to ~1 hour
- Updated service query logic for Windows
- Updated golang sys package dependency

</p>
</details>

<details><summary>3.0.755.0</summary>
<p>

- Fix fallback logic for MGS endpoint generation
- Fix regional endpoint generation

</p>
</details>

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

jpculp committed:
Add support for hybrid environments

This change allows users running on-premises or in a hybrid environment
to register with SSM as a managed-instance. SSM's local state directory
(/var/lib/amazon/ssm) has moved to a persistent directory in
/.bottlerocket/host-containers/current/. Activations parameters are
parsed from a base64-encoded JSON block in the user-data setting.

Example JSON:
{
  "ssm":{
    "activation-id":"foo",
    "activation-code":"bar",
    "region":"us-west-2",
  }
}

webern commented on Dockerfile:
```diff
@@ -22,16 +22,18 @@ LABEL "org.opencontainers.image.version"="$IMAGE_VERSION"
-  rm -rf /var/cache/yum ./hashes
+  rm -rf /var/cache/yum ./hashes && \
+  rmdir /var/lib/amazon/ssm && \
+  ln -snf /.bottlerocket/host-containers/control/ssm /var/lib/amazon/ssm
```
Maybe explain that `/.bottlerocket/host-containers/control` will be mounted when it runs on Bottlerocket? Dunno, maybe that's obvious.

Is this the first time that the control container becomes sensitive to it having the exact name `control` in Bottlerocket. i.e. before this I could have used this container and called it host-containers.foo on Bottlerocket and it would have worked? Does a host container have a way of finding out its name at runtime, and if so maybe we could create the link at runtime?

I'm kind of surprised that we mount host container paths in this non-abstracted way to begin with. I would have thought we would mount `/.bottlerocket/host-containers/control/` as `bottlerocket` for a host container or something like that to decouple things, but I guess I wasn't paying attention to that.

tjkirch replied:
Yeah, making ourselves sensitive to host container name would be unfortunate.  Do we have another way to get this name, or should we change the mounting strategy, or create some other way to use a consistent path instead of having to know the name..?

jpculp replied:
My guess is that if someone is using a custom name for their control container, they are probably sporting a custom Dockerfile as well. I can certainly add a comment as a heads up. Currently the admin container logic is also sensitive to `/.bottlerocket/host-containers/admin/`. If we go the route of changing how it mounts here then we'll want to change it there as well to remain consistent.

tjkirch replied:
What if you want "control" to be your custom container, and "control-br" to be the official one?  I'm sure there are other similar cases.  There's no need to tie ourselves to a given name such that it'd break when the user customizes something we let them customize.

zmrow replied:
> ... Do we have another way to get this name, or should we change the mounting strategy, or create some other way to use a consistent path instead of having to know the name..?

This is where my mind went first.  Where are we going to get the name?  Off-the-cuff:  we could potentially pass the name into the container at runtime via the env file?  Just spitballing here...

jpculp replied:
To address this, I've put out a PR on bottlerocket [#1416](https://github.com/bottlerocket-os/bottlerocket/pull/1416). By adding a second bind mount with a more generic name we can more safely reference the location in scripts.

webern commented on start_control_ssm.sh:
```diff
@@ -0,0 +1,64 @@
+  region=$(get_user_data_ssm "region")
+
+  # If any of the parameters are missing, then exit 1.
+  if [[ "${failed_ssm_params}" -ne 0 ]]; then
```
Maybe I'm bad at shell, but how could this ever be anything but zero? Did you mean to check an exit code or something?

jpculp replied:
The logic is a bit nested, but it's incremented in `get_user_data_ssm`. Because of that I didn't make it a local variable. This was out of a desire to not rely on `set -e`.

zmrow replied:
I'd rather not create a variable here inside a function that gets treated like a contstant (even though bash is fine with this).  I think rather than keeping a count, we can assume that all parameters are needed and fail if any of them aren't found.

webern commented on start_control_ssm.sh:
```diff
@@ -0,0 +1,64 @@
+  local ssm_param="${1:?}"
+  if ! param_value=$(jq --arg ssm_param "${ssm_param}" -e -r '.["ssm"][$ssm_param]' "${USER_DATA}"); then
+    log "Failed to parse ${ssm_param} from ${USER_DATA}"
+    ((++failed_ssm_params))
```
Oh I see, sneaky.

tjkirch replied:
Yeah, too sneaky - can we return nonzero instead, and handle the counting in the other function?

jpculp replied:
Rather than count, we can just exit 1. The counting was more of a hold over from the original design where we weren't going to exit if SSM failed to register.

zmrow commented on README.md:
```diff
@@ -9,3 +9,29 @@ For more information about the control container, including how to use it and ho
+
+## Connecting to AWS Systems Manager (SSM)
+
+Starting from v0.5.0, users have the option to pass in their own activations for SSM. This is for users that want to set up on-premises virtual machines (VMs) in their hybrid environment as managed instances.
```
Our convention has been to write markdown files with one sentence per line.  If there are no empty lines, they get formatted as a paragraph.  If you want a new paragraph, leave a blank line.  It just makes these docs a bit easier to read/edit.

zmrow commented on README.md:
```diff
@@ -9,3 +9,29 @@ For more information about the control container, including how to use it and ho
+
+## Connecting to AWS Systems Manager (SSM)
+
+Starting from v0.5.0, users have the option to pass in their own activations for SSM. This is for users that want to set up on-premises virtual machines (VMs) in their hybrid environment as managed instances.
```
Should this say "pass in their own activation information"?

jpculp replied:
I'm not sure actually. The AWS documentation refers to them as Hybrid Activations, but both I think could be correct.

zmrow commented on README.md:
```diff
@@ -9,3 +9,29 @@ For more information about the control container, including how to use it and ho
+
+Starting from v0.5.0, users have the option to pass in their own activations for SSM. This is for users that want to set up on-premises virtual machines (VMs) in their hybrid environment as managed instances.
+
+Users can add their [own activations](https://docs.aws.amazon.com/systems-manager/latest/userguide/sysman-managed-instance-activation.html) by populating the control container's user-data with a base64-encoded JSON block.
```
Same here.

Also I don't think we need the hyphen in "user-data" or "base64-encoded".  Editor @tjkirch might throw something at me... :)

tjkirch replied:
The EC2 standard in prose is "user data", with a space and no hyphen.  "base64-encoded" should be hyphenated when used as an adjective like it is here.  :)

jpculp replied:
Up to this point I've been using "user-data" to make it more clear that this is "host-containers.user-data" and not EC2's "user data". I take it I should just refer to both as user data going forward?

tjkirch replied:
In the specific case referred to here, I'd say it's prose, since we're still in the "introduction", talking about the user data belonging to a container and not exactly describing how to set it with a setting that happens to be called `user-data`.  If you added the word "setting" afterward, then it'd be more specific, but I don't think we're to the "here's exactly how you do this" phase in the doc quite yet.

Down below, where it's saying the control container's "user-data setting" I'd go with the hyphen, plus backticks, to make it clear that now we're talking about a technical thing, a specific setting name, `user-data`.  (You also have the example TOML, which makes it clear that a hyphen is needed for that format.)

zmrow commented on README.md:
````diff
@@ -9,3 +9,29 @@ For more information about the control container, including how to use it and ho
+
+```
+{
+  "ssm": {
````
Where would one find this information?  Can we point them to the top level docs for this?

zmrow commented on README.md:
````diff
@@ -9,3 +9,29 @@ For more information about the control container, including how to use it and ho
+}
+```
+
+Once you've created your JSON, you'll need to base64-encode it and set it as the value of the control host container's user-data setting in your [instance user data toml](https://github.com/bottlerocket-os/bottlerocket#using-user-data).
````
Same here - I don't think we need the hyphens.

zmrow commented on start_control_ssm.sh:
```diff
@@ -0,0 +1,64 @@
+  # to manually register with SSM as a hybrid environment host.
+  declare -r PERSISTENT_CONTROL_DIR="/.bottlerocket/host-containers/control"
+  declare -r SSM_AGENT_LOCAL_STATE_DIR="/var/lib/amazon/ssm"
+  declare -r SSM_AGENT_PERSISTENT_STATE_DIR="${PERSISTENT_CONTROL_DIR}/ssm"
```
Super nit: Can we bump this variable up one line so it's directly under the variable it's using (PERSISTENT_CONTROL_DIR)?  IMHO, makes it a tiny bit easier to follow the variable chaining.

zmrow commented on start_control_ssm.sh:
```diff
@@ -0,0 +1,64 @@
+  declare -r PERSISTENT_CONTROL_DIR="/.bottlerocket/host-containers/control"
+  declare -r SSM_AGENT_LOCAL_STATE_DIR="/var/lib/amazon/ssm"
+  declare -r SSM_AGENT_PERSISTENT_STATE_DIR="${PERSISTENT_CONTROL_DIR}/ssm"
+  declare -r USER_DATA="${PERSISTENT_CONTROL_DIR}/user-data"
```
Same here. :)

zmrow commented on start_control_ssm.sh:
```diff
@@ -0,0 +1,64 @@
+  echo "${param_value}"
+}
+
+main() {
```
Unless someone really wants this format, I'd like to eliminate a `main()` function and un-nest the logic below.  I don't see it as necessary in a script of this size.

jpculp commented:
- Replaced all instances of `/.bottlerocket/host-containers/control` with `/.bottlerocket/host-containers/current` ("current" directory introduced in bottlerocket [#1416](https://github.com/bottlerocket-os/bottlerocket/pull/1416)).
- Removed `failed_ssm_params` and replaced error handling with `exit 1`.
- Removed explicit `main` function.
- Shuffled some constants around for easier reading.
- Small tweaks in `README` and comments.

tjkirch commented on CHANGELOG.md:
```diff
@@ -1,3 +1,11 @@
+* Symlinked `/var/lib/amazon/ssm` to `/.bottlerocket/host-containers/current/ssm` so that SSM Agent state data can persist between boots. ([#12])
+* Update SSM Agent version to 3.0.755.0. ([#12])
+
+[#12]: https://github.com/bottlerocket-os/bottlerocket-control-container/pull/
```
```suggestion
[#12]: https://github.com/bottlerocket-os/bottlerocket-control-container/pull/12
```

tjkirch commented on start_control_ssm.sh:
```diff
@@ -0,0 +1,52 @@
+if [[ -s "${USER_DATA}" ]] && [[ ! -s "${SSM_AGENT_LOCAL_STATE_DIR}/registration" ]]; then
+  if jq -e -r '.["ssm"]' "${USER_DATA}" &>/dev/null; then
+    enable_hybrid_env_ssm
+  fi
```
```suggestion
if [[ -s "${USER_DATA}" ]] \
&& [[ ! -s "${SSM_AGENT_LOCAL_STATE_DIR}/registration" ]] \
&& jq --exit-status '.ssm' "${USER_DATA}" &>/dev/null \
; then
  enable_hybrid_env_ssm
```
* nit on layout to make the list of conditions clearer
* nit on spelling out `-e` so we don't have to look it up as often
* nit on simplifying jq condition
* don't think we need `-r` since we're throwing out the output anyway

tjkirch commented on start_control_ssm.sh:
```diff
@@ -0,0 +1,52 @@
+get_user_data_ssm() {
+  local param_value
+  local ssm_param="${1:?}"
+  if ! param_value=$(jq --arg ssm_param "${ssm_param}" -e -r '.["ssm"][$ssm_param]' "${USER_DATA}"); then
```
```suggestion
  if ! param_value=$(jq --arg ssm_param "${ssm_param}" -e -r '.ssm[$ssm_param]' "${USER_DATA}") \
  || [ -z "${param_value}" ]; then
```
I think we always need non-empty values...  (and a little simplification of the condition)

tjkirch commented on start_control_ssm.sh:
```diff
@@ -0,0 +1,52 @@
+# Extract the values from user-data json, and exit on failure (if any)
+get_user_data_ssm() {
+  local param_value
+  local ssm_param="${1:?}"
```
Might be good to take the file as a parameter too, rather than assuming a global USER_DATA, so that the function is easier to understand on its own.

I'm not sure this ssm_param argument is carrying its weight - it seems simpler and clearer to me to have a general `get_json` (or whatever) function that you pass `.ssm.activation-code` instead of assuming the `.ssm` prefix and doing the jq variable magic to stitch them together.

tjkirch commented on start_control_ssm.sh:
```diff
@@ -0,0 +1,52 @@
+
+  # Register with AWS Systems Manager (SSM)
+  amazon-ssm-agent -register -code "${activation_code}" -id "${activation_id}" -region "${region}"
+  if [[ "$?" -eq 1 ]]; then
```
```suggestion
  if [[ "$?" -ne 0 ]]; then
```
We don't want to have to care if they start returning 3 for some errors :)

tjkirch commented on start_control_ssm.sh:
```diff
@@ -0,0 +1,52 @@
+  # Register with AWS Systems Manager (SSM)
+  amazon-ssm-agent -register -code "${activation_code}" -id "${activation_id}" -region "${region}"
+  if [[ "$?" -eq 1 ]]; then
+    log "Failed to register with AWS Systems Manager (SSM)"
```
Would you please add some more testing data about failure cases?  I'm particularly interested in how hard it is to troubleshoot when this goes wrong; do these messages show up in the EC2 console?

tjkirch commented on README.md:
````diff
@@ -9,3 +9,30 @@ For more information about the control container, including how to use it and ho
+}
+```
+
+Once you've created your JSON, you'll need to base64-encode it and set it as the value of the control host container's `user-data` setting in your [instance user data toml](https://github.com/bottlerocket-os/bottlerocket#using-user-data).
````
```suggestion
Once you've created your JSON, you'll need to base64-encode it and put it in the control container's `user-data` setting in your [instance user data](https://github.com/bottlerocket-os/bottlerocket#using-user-data).
For example:
```

tjkirch replied:
That sentence is still complex, so I do think the friendly "For example:" afterward helps :)

tjkirch commented on Dockerfile:
```diff
@@ -22,16 +22,18 @@ LABEL "org.opencontainers.image.version"="$IMAGE_VERSION"
-  rm -rf /var/cache/yum ./hashes
+  rm -rf /var/cache/yum ./hashes && \
+  rmdir /var/lib/amazon/ssm && \
+  ln -snf /.bottlerocket/host-containers/current/ssm /var/lib/amazon/ssm
```
So we don't forget: if we change the approach in https://github.com/bottlerocket-os/bottlerocket/pull/1416 to use a variable (or something) we'll have to change the approach here, perhaps even dynamically creating the link in start_control_ssm.sh.

jpculp commented:
- Moved some logic back out of the Dockerfile to a function in `start_control_ssm.sh`.
- Replaced `get_user_data_ssm` with a more generic `fetch_from_json` which takes the `user-data` file as the 2nd argument.
- Replaced "current" with environment variable `HOST_CONTAINER_NAME`, which will get set by `host-ctr`.
- Added error handling logic if `jq` returns an empty value.
- Small tweaks to enhance readability of conditionals leading to `enable_hybrid_ssm`.
- Small tweaks to `README`.

jpculp commented on CHANGELOG.md:
```diff
@@ -1,3 +1,11 @@
+# 0.5.0
+
+* Add support for on-premises and hybrid environments. ([#12])
+* Symlinked `/var/lib/amazon/ssm` to `/.bottlerocket/host-containers/NAME/ssm` so that SSM Agent state data can persist between boots. ([#12])
```
Does anyone have a preference on if this should be `/NAME/`, `/<name>/`, or `/${HOST_CONTAINER_NAME}/`?

webern replied:
Isn't it going to be `current`?

jpculp replied:
We're going to go with an environment variable instead. Check out https://github.com/bottlerocket-os/bottlerocket/pull/1422.

tjkirch replied:
We should probably use the real variable here, right?  It's still as clear when reading, and can actually be used.

webern commented on start_control_ssm.sh:
```diff
@@ -0,0 +1,67 @@
+# and the symlinked /var/lib/amazon/ssm/registration file is not populated,
+# then check to see if the user-data file contains ssm at the top-level. If so,
+# attempt to manually register with SSM with a hybrid activation.
+declare -r PERSISTENT_STORAGE_BASE_DIR="/.bottlerocket/host-containers/${HOST_CONTAINER_NAME}"
```
Wasn't this why we made the change in Bottlerocket `host-ctr`?

```suggestion
declare -r PERSISTENT_STORAGE_BASE_DIR="/.bottlerocket/host-containers/current"
```

tjkirch commented on start_control_ssm.sh:
```diff
@@ -0,0 +1,67 @@
+
+if [[ -s "${USER_DATA}" ]] \
+&& [[ ! -s "${SSM_AGENT_LOCAL_STATE_DIR}/registration" ]] \
+&& jq --exit-status '.ssm' "${USER_DATA}" &>/dev/null ; then
```
This is basically `fetch_from_json`; we'd probably want to fail here, too, if `.ssm` was empty inside.

I think it might be easier to use `set -e` for the whole script.  Then you could use `fetch_from_json` here too.  I think then you could remove the `set -e` in `prepare_persistent_state_dir`, make `fetch_from_json` return 1 instead of exit 1, and change the main call to `if ! amazon-ssm-agent -bla` instead of checking `$?` afterward, and it'd be fine?  If there are any lines we expect could fail, we could use `if ! bla; log explanation; exit 1; fi` for extra clarity, but I wouldn't expect the simple mkdir/chmod/rmdir/ln to fail.

tjkirch commented on start_control_ssm.sh:
```diff
@@ -0,0 +1,67 @@
+  local key="${1:?}"
+  local file="${2:?}"
+  local value
+  if ! value=$(jq -e -r '$key' "${file}") || [[ -z "${value}" ]]; then
```
Hmm - it seems like it should be `"$key"` rather than `'$key'`?  The shell wouldn't replace `$key` inside single quotes, does jq do that?!  (Even if it does, I think it's unexpected when reading shell, so I'd use double...)

jpculp replied:
It should be "${key}" instead of '$key' since we removed the --arg . Sorry about that.

jpculp commented:
- Removed reliance on a `HOST_CONTAINER_NAME` environment variable and instead hard-coded to `/.bottlerocket/host-containers/current` which was added in [bottlerocket #1416](https://github.com/bottlerocket-os/bottlerocket/pull/1416)
- Fixed quoting issues when fetching data from `jq`
- Add `set -e` to the top of the script
- Replaced `$?` return code check with an `if !` for `amazon-ssm-agent -register` 
- Replaced single quotes with double quotes around `key` variable
- Split `prepare_persistent_state_dir` function back into two pieces (Dockerfile and script)
- Bumped SSM version from 3.0.755.0 to 3.0.882.0

tjkirch commented on start_control_ssm.sh:
```diff
@@ -0,0 +1,54 @@
+
+enable_hybrid_env_ssm() {
+  # SSM parameters necessary to register with a hybrid activation
+  activation_code=$(fetch_from_json '."ssm"."activation-code"' "${USER_DATA}")
+  activation_id=$(fetch_from_json '."ssm"."activation-id"' "${USER_DATA}")
+  region=$(fetch_from_json '."ssm"."region"' "${USER_DATA}")
```
```suggestion
  activation_code=$(fetch_from_json '.ssm."activation-code"' "${USER_DATA}")
  activation_id=$(fetch_from_json '.ssm."activation-id"' "${USER_DATA}")
  region=$(fetch_from_json '.ssm.region' "${USER_DATA}")
```
nit: turns out it's the hyphen that needs quoting.  I'd vote for using the simpler unquoted form where possible, and just quoting the hyphenated parts, to make it clearer that that's why the extra quotes are there.

Another option is square bracketing everything to make it consistent, with the quotes standing out a little less (e.g. `.["ssm"]["activation-id"]`) but since plain dots are most common I'd vote to use them where possible, and quote where not, but it's not a big deal.

jpculp replied:
I would prefer `.["ssm"]["activation-id"]` to `'.ssm."activation-id"'` or `'."ssm"."activation-id"'` since it is consistent with the admin container and also seems less fragile to tweaks (ie. someone changes the query slightly x years from now), if you are ok with it.

tjkirch reviewed (commented):
The new changes look good to me.  The main thing I'm hoping to see is more testing of failure cases - https://github.com/bottlerocket-os/bottlerocket-control-container/pull/12#discussion_r600955507.  I want to make sure it's going to be possible to troubleshoot issues with activations using console output.  (Also a couple nit comments open.)

jpculp commented:
- Replaced `'."ssm"."activation-code"'` format with `.["ssm"]["activation-code"]` for readability, maintainability, and consistency with the admin container
- Rather than print the last line of the error file, read the whole file line-by-line.

For some context on the last piece, @tjkirch and I were concerned that if we printed the whole log file it might print errors from previous boots. In my testing, I found this not to be the case as the log file was fresh each time.

jpculp commented:
Updated the PR description to include the output of several failure cases. _(Click each case to reveal EC2 console output)_

zmrow reviewed (approved):
Looking good! :jigsaw:

jpculp commented:
Updated the PR description to include the [SSM Agent changes](https://github.com/aws/amazon-ssm-agent/releases) since 3.0.732.0 (the version we included in v0.4.2)

jahkeup commented on README.md:
````diff
@@ -9,3 +9,32 @@ For more information about the control container, including how to use it and ho
+```
+{
+  "ssm": {
+    "activation-id": "foo",
+    "activation-code": "bar",
+    "region":"us-west-2"
````
These names aren't consistent with the user data used in the admin container (`foo-bar` vs `foo_bar`). Was this intentional? As much as we can, I'd prefer that the "convention" stays consistent for users. This aligns with Bottlerocket's `user-data.toml` naming (which uses skewers) so one *could* argue that this one is the "right way" and the admin is in the wrong.. but we already have the admin change out, so I'm more inclined to align with that user data as this user data and the admin's, to me, roughly feels like it'll be treated as same namespace and with similar conventions by end users.

jpculp replied:
Based on the other settings I see when I run `apiclient -u /settings` I would say that the admin container is the weird one, and that `foo-bar` is the "correct" convention, but I do not have strong feelings. I think the it might be worth getting some additional opinions. I think the admin container is the way it is because the authorized keys are typically stored in a file named `authorized_keys`, but that may have been a poor decision on my part. To make this conundrum even more fun, when you generate an activation token with `aws ssm create-activation` it returns JSON with a `FooBar` convention.

jpculp replied:
After some offline discussion, we're going to keep the admin container convention and use `foo_bar` (snake_case) for the settings inside host-container user data in general. I'll add this change.

jhaynes replied:
I'm in favor of sticking with the skewer (`foo-bar`) implementation here and adding a backlog item to fix the admin container to accept the skewer also.

jahkeup commented on start_control_ssm.sh:
```diff
@@ -0,0 +1,60 @@
+
+enable_hybrid_env_ssm() {
+  # SSM parameters necessary to register with a hybrid activation
+  activation_code=$(fetch_from_json '.["ssm"]["activation-code"]' "${USER_DATA}")
+  activation_id=$(fetch_from_json '.["ssm"]["activation-id"]' "${USER_DATA}")
+  region=$(fetch_from_json '.["ssm"]["region"]' "${USER_DATA}")
```
Please make scoped variables, well, scoped!

```suggestion
  local activation_code=$(fetch_from_json '.["ssm"]["activation-code"]' "${USER_DATA}")
  local activation_id=$(fetch_from_json '.["ssm"]["activation-id"]' "${USER_DATA}")
  local region=$(fetch_from_json '.["ssm"]["region"]' "${USER_DATA}")
```

jpculp replied:
According to [SC2155](https://github.com/koalaman/shellcheck/wiki/SC2155) this can mask return values. How do you feel about:
```
  local activation_code activation_id region
  activation_code=$(fetch_from_json '.["ssm"]["activation-code"]' "${USER_DATA}")
  activation_id=$(fetch_from_json '.["ssm"]["activation-id"]' "${USER_DATA}")
  region=$(fetch_from_json '.["ssm"]["region"]' "${USER_DATA}")
  ```

jahkeup commented on start_control_ssm.sh:
```diff
@@ -0,0 +1,60 @@
+
+    # Print errors line-by-line from ssm agent error log,
+    # as they don't print to the EC2 console otherwise.
+    while IFS= read -r line; do
+      log "${line}"
+    done < "/var/log/amazon/ssm/errors.log"
```
Does printing the log file directly to the console (without chunking lines) cause the output to become garbled?

I'd prefer to see this written with a simpler, single command rather than iteration:

```bash
cat /var/log/amazon/ssm/errors.log >&2
```

In fact, as is (and with the above) the log file would grow after each attempt and keep on going.. so really, we should truncate the file when attempting to register. To keep around some of the older logs, we can do a cheap "logrotate" and rename the existing files to allow for debugging while allowing SSM to write its errors to the fresh log. This would ensure that there wasn't **too much printed** to the console (making `cat` a practical choice) and that the size of the log doesn't become overwhelming over time.

jpculp replied:
I can't say I've seen it garbled personally, so it may not be necessary. I'm definitely keen on using a simple `cat` if nobody has concerns.

Regarding truncating the file, from what I saw in my testing the errors.log did not survive between attempts, most likely since it's not part of the persistent storage. That said, if folks would feel more comfortable with a logrotate I can certainly add it.

jpculp commented:
- Switched from kebab-case to snake_case for `user-data` settings
- Moved constants to the top
- Set the scope of `enable_hybrid_env_ssm` vars to local
- Replaced line-by-line error log print with a single cat
- Split up the `value` jq logic

jahkeup commented on Dockerfile:
```diff
@@ -52,4 +54,6 @@ RUN chmod +x /usr/bin/enable-admin-container
 RUN useradd -m -G users,api ssm-user
 
-CMD ["/usr/bin/amazon-ssm-agent"]
+ADD start_control_ssm.sh /usr/sbin/
+RUN chmod +x /usr/sbin/start_control_ssm.sh
```
If you set the bits on the file that's added, then it comes in with the executable bits set - which avoids this extra command!

jpculp replied:
Rather than change that in this PR, I've opened up an issue ([bottlerocket-os/bottlerocket/issues/1438](https://github.com/bottlerocket-os/bottlerocket/issues/1438)). I found a neat solution, but it requires Docker 20.10.

jahkeup commented on start_control_ssm.sh:
```diff
@@ -0,0 +1,68 @@
+#!/bin/bash
```
Not needed.. but I do prefer we use the portable convention everywhere that it is possible to:

```suggestion
#!/usr/bin/env bash
```

jahkeup reviewed (changes requested):
Please set the permission bits on the script before adding - and then.. ship it!

🚢 :shipit: 👍🏽

jpculp committed:
Update SSM agent to 3.0.882.0

jpculp committed:
Version v0.5.0

jpculp commented:
- Switched `#!/bin/bash` to `#!/usr/bin/env bash`
- Reverted back to kebab-case as it is our API standard (after further offline discussion)

jahkeup commented on README.md:
````diff
@@ -9,3 +9,32 @@ For more information about the control container, including how to use it and ho
+}
+```
+
+Once you've created your JSON, you'll need to base64-encode it and put it in the control host container's `user-data` setting in your [instance user data](https://github.com/bottlerocket-os/bottlerocket#using-user-data).
+
+For example:
````
I took a stab at rewording this because it took a pass or two to full parse the sentence. Now that its split up I think this conveys the same and reads a bit easier (IMO) - what do you think about this instead?

```suggestion
Once you've created your JSON, you'll need to provide it via the control container's user data.
The host-container `user-data` must be base64-encoded and provided - in its encoded form - using your [instance's user data](https://github.com/bottlerocket-os/bottlerocket#using-user-data):
```

jpculp replied:
I think I prefer the original, but I appreciate that you're looking at the documentation 😃

jahkeup reviewed (approved):
Have a suggest edit for the README - script looks great!

bcressey commented on README.md:
````diff
@@ -9,3 +9,32 @@ For more information about the control container, including how to use it and ho
+
+```
+[settings.host-containers.control]
+# ex: echo '{"ssm":{"activation-id":"foo","activation-code":"bar","region":"us-west-2"}}' | base64
````
On my system, I need `base64 -w0` to avoid newlines in the output.
```suggestion
# ex: echo '{"ssm":{"activation-id":"foo","activation-code":"bar","region":"us-west-2"}}' | base64 -w0
```

jpculp replied:
I think we avoided it since not all systems may necessarily have `-w`. (MacOS, for example, does not)

jpculp commented:
Ran the tests again both with and without a hybrid activation in `user-data` using an AMI built off the latest develop branch. Still all green!

zmrow reviewed (approved):
:waffle:
